src/lib/ptree/ptree.lua

```diff
@@ -527,6 +542,63 @@ end
 
 local dummy_unix_sockaddr = S.t.sockaddr_un()
 
+function Manager:push_notifications_to_peers()
+   local notifications = alarms.notifications()
+   if #notifications == 0 then return end
+   local function head (queue)
+      local msg = assert(queue[1])
+      local len = #msg
+      local buf = ffi.new("char[?]", len)
+      ffi.copy(buf, msg, len)
+      return buf, len
+   end
+   local function tojson (output, str)
+      json_lib.write_json_object(output, str)
+      local msg = output:flush()
+      return tostring(#msg)..'\n'..msg
+   end
+   -- Enqueue notifications into each peer queue.
+   local peers = self.notification_peers
+   for _,peer in ipairs(peers) do
+      local output = json_lib.buffered_output()
+      peer.queue = peer.queue or {}
+      for _,each in ipairs(notifications) do
+         table.insert(peer.queue, tojson(output, each))
+      end
+   end
+   -- Iterate peers and send enqueued messages.
+   for i,peer in ipairs(peers) do
+      local queue = peer.queue
+      while #queue > 0 do
+         local buf, len = head(peer.queue)
+         peer.pos = peer.pos or 0
+         local count, err = peer.fd:write(buf + peer.pos,
+                                          len - peer.pos)
```

Comment:
You can actually do `ffi.cast('uint8_t*', queue[1]) + peer.pos` here to avoid the allocation in `head`.

Reply:
This is fixed although I changed `head` to return a pointer instead of doing an allocation, as suggested.